manticore/core/smtlib/visitors.py

```diff
@@ -232,29 +253,35 @@ class ArithmeticSimplifier(Visitor):
     def __init__(self, parent=None, **kw):
         super(ArithmeticSimplifier, self).__init__(**kw)
 
-    def _method(self, expression, *operands):
-        value = super(ArithmeticSimplifier, self)._method(expression, *operands)
-        #while value is not expression:
-        #    expression = value
-        #    if isinstance(expression, Operation):
-        #        print "A", expression.operands
-        #        operands = [self._method(op, *op.operands) for op in expression.operands]
-        #        print "B", operands
-        #        value = super(ArithmeticSimplifier, self)._method(expression, *operands)
-        #something changed recursively visit the new node.
-        if expression is not value:
-            self.visit(value)
-            value = self.pop()
-
-
-        #if value is not expression
-        #    if isinstance(value, Operation):
-        #        for i in xrange(len(value.operands)):
-        #            self.visit(op)
-        #        new_operands = reversed([self.pop() for _ in range(len(value.operands))])
-        #        value = super(ArithmeticSimplifier, self)._method(value, *new_operands)
-        return value
-                   
+    def visit(self, node):
+        cache = self._cache
+
+        visited = set()
+        stack = []
+        stack.append(node)
+        while stack:
+            node = stack.pop()
+ 
+            if node in cache:
+                self.push(cache[node])
+            elif isinstance(node, Operation):
+                if node in visited:
+                    operands = [self.pop() for _ in xrange(len(node.operands))]
+                    new_node = self._rebuild(node, operands)
+                    value = self._method(new_node, *operands)
+                    while value is not new_node:
```

Comment:
I wonder if we can have a generic  visit() in the Visitor class that can account for this cases. Basically If we can move this thing up in the class hierarchy. Maybe overload _rebuild() and set a class configuration flag to control if i needs to re-runn ._method() until it finds a fixed point?